File: examples/global-ot-chat/src/main/java/io/global/ot/chat/operations/ChatOTSystem.java
package io.global.ot.chat.operations;

import io.datakernel.ot.OTSystem;
import io.datakernel.ot.OTSystemImpl;
import io.datakernel.ot.TransformResult;
import io.global.ot.chat.common.Operation;

import static java.util.Collections.singletonList;

public class ChatOTSystem {

	public static OTSystem<Operation> createOTSystem() {
		return OTSystemImpl.<Operation>create()
				.withEmptyPredicate(ChatOperation.class, operation -> operation.getContent().isEmpty())
				.withInvertFunction(ChatOperation.class, op -> singletonList(op.invert()))
				.withTransformFunction(ChatOperation.class, ChatOperation.class, (left, right) -> TransformResult.of(right, left));
	}
}

File: examples/global-ot-chat/front/src/modules/chat/ot/chatOTSystem.js
import {OTSystemBuilder, TransformResult} from 'ot-core';
import ChatOTOperation from './ChatOTOperation';

const chatOTSystem = new OTSystemBuilder()
  .withEmptyPredicate(ChatOTOperation, operation => operation.isEmpty())
  .withInvertFunction(ChatOTOperation, operation => operation.invert())
  .withSquashFunction(ChatOTOperation, ChatOTOperation, (operationLeft, operationRight) => {
    if (operationLeft.isEmpty()) {
      return operationRight;
    }

    if (operationRight.isEmpty()) {
      return operationLeft;
    }

    if (operationRight.isEqual(operationLeft.invert())) {
      return ChatOTOperation.EMPTY;
    }
  })
  .withTransformFunction(ChatOTOperation, ChatOTOperation, (operationLeft, operationRight) => {
    return TransformResult.of(operationRight, operationLeft);
  })
  .build();

export default chatOTSystem;
